core/sql/parser/sqlparser.y

```diff
@@ -6647,15 +6651,26 @@ table_as_tmudf_function : TOK_UDF '(' table_mapping_function_invocation ')'
 
 table_name_and_hint : table_name optimizer_hint hbase_access_options
                       {
-                        $$ = new (PARSERHEAP()) Scan(*$1);
-                        if ($2) 
-                          $$->setHint($2);
+                        NAString tmp = ((*$1).getQualifiedNameAsString());
+                        if(SqlParser_CurrentParser->hasWithDefinition(&tmp) )
+                        {
+                          RelExpr *re = SqlParser_CurrentParser->getWithDefinition(&tmp);
+                          $$=re->copyTree(PARSERHEAP());
+                          delete $1;
```

Comment:
One thing we could do here is to raise an error if we have an optimizer hint or HBase access options at this point. Those are allowed in the grammar, but we wouldn't be able to apply them to a more complicated WITH clause.